MantidPlot/src/ProjectSerialiser.cpp

```diff
@@ -0,0 +1,946 @@
+  // If we're not the top level folder, read the folder settings and create the
+  // folder
+  // This is a legacy edgecase because folders are written
+  // <folder>\tsettings\tgo\there
+  if (!isTopLevel && lines.size() > 0) {
+    std::vector<std::string> lineVec;
+    boost::split(lineVec, lines, boost::is_any_of("\n"));
+
+    std::string firstLine = lineVec.front();
+
+    std::vector<std::string> values;
+    boost::split(values, firstLine, boost::is_any_of("\t"));
+
+    auto *newFolder =
+        new Folder(window->currentFolder(), QString::fromStdString(values[1]));
+    newFolder->setBirthDate(QString::fromStdString(values[2]));
+    newFolder->setModificationDate(QString::fromStdString(values[3]));
+
+    if (values.size() > 4 && values[4] == "current")
+      window->d_loaded_current = newFolder;
+
+    auto *fli = new FolderListItem(
+        window->currentFolder()->folderListItem(), newFolder);
+    newFolder->setFolderListItem(fli);
+
+    window->d_current_folder = newFolder;
+
+    // Remove the first line (i.e. the folder's settings line)
+    lineVec.erase(lineVec.begin());
+    lines = boost::algorithm::join(lineVec, "\n");
+  }
+
+  loadProjectSections(lines, fileVersion, isTopLevel);
+
+  // We're returning to our parent folder, so set d_current_folder to our parent
+  auto *parent = dynamic_cast<Folder *>(window->currentFolder()->parent());
+  if (!parent)
+    window->d_current_folder = window->projectFolder();
+  else
+    window->d_current_folder = parent;
+}
+
+/**
+ * Load sections of the project file back into Mantid
+ *
+ * This function looks at individual sections of the TSV project file
+ * and loads the relevant windows etc.
+ *
+ * @param lines :: string of characters from a Mantid project file.
+ * @param fileVersion :: version of the project file loaded
+ * @param isTopLevel :: whether this is being called on a top level folder.
+ */
+void ProjectSerialiser::loadProjectSections(const std::string &lines,
+                                            const int fileVersion,
+                                            const bool isTopLevel) {
+  // This now ought to be the regular contents of a folder. Parse as normal.
+  TSVSerialiser tsv(lines);
+
+  // If this is the top level folder of the project, we'll need to load the
+  // workspaces before anything else.
+  if (isTopLevel && tsv.hasSection("mantidworkspaces")) {
+    // There should only be one of these, so we only read the first.
+    std::string workspaces = tsv.sections("mantidworkspaces").front();
+    populateMantidTreeWidget(QString::fromStdString(workspaces));
+  }
+
+  if (tsv.hasSection("open")) {
+    std::string openStr = tsv.sections("open").front();
+    int openValue = 0;
+    std::stringstream(openStr) >> openValue;
+    window->currentFolder()->folderListItem()->setExpanded(openValue);
+  }
+
+  if (tsv.hasSection("mantidmatrix")) {
+    auto matrices = tsv.sections("mantidmatrix");
+    for (auto it : matrices) {
```

Comment:
We should be taking the elements by reference, i.e. `for (auto& it : matrices) ...`. Refrences (and cv-qualifieres) need to be specified explicitly, since the compiler cannot tell if you want to grab each element via reference or value. As it is now, we will have a performance hit, since every element is copied into the local variable.  

Sorry, I should have mentioned that earlier on.